# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsIGlobalObject.cpp
# Commit: 5278e40b80c3
# Full Hash: 5278e40b80c3c4aa642a24a1ff5a50247b9b3126
# Author: Tom Ritter <tom@mozilla.com>
# Date: 2022-11-28 15:41:50
# Regressor Bug: 1778510
# File Overlap Count: 1
# Description:
#   Bug 1778510: Add RTPCallerType to GlobalObject r=asuth
#   
#   This centralizes the logic in one place.
#   
#   In order to do this, we will need to check the principal
# ==============================================================================

diff -r dba220581d93 -r 5278e40b80c3 dom/base/nsIGlobalObject.cpp
--- a/dom/base/nsIGlobalObject.cpp	Mon Nov 28 04:21:26 2022 +0000
+++ b/dom/base/nsIGlobalObject.cpp	Mon Nov 28 04:21:26 2022 +0000
@@ -68,7 +68,7 @@
   MOZ_DIAGNOSTIC_ASSERT(mEventTargetObjects.isEmpty());
 }
 
-nsIPrincipal* nsIGlobalObject::PrincipalOrNull() {
+nsIPrincipal* nsIGlobalObject::PrincipalOrNull() const {
   if (!NS_IsMainThread()) {
     return nullptr;
   }
@@ -396,3 +396,26 @@
 
   return mozilla::ipc::StorageKeysEqual(storageKey, aStorageKey);
 }
+
+bool nsIGlobalObject::IsSystemPrincipal() const {
+  MOZ_ASSERT(NS_IsMainThread(),
+             "Cannot ask nsIGlobalObject IsSystemPrincipal off-main-thread");
+
+  return PrincipalOrNull()->IsSystemPrincipal();
+}
+
+RTPCallerType nsIGlobalObject::RTPCallerType() const {
+  if (IsSystemPrincipal()) {
+    return RTPCallerType::SystemPrincipal;
+  }
+
+  if (ShouldResistFingerprinting()) {
+    return RTPCallerType::ResistFingerprinting;
+  }
+
+  if (CrossOriginIsolated()) {
+    return RTPCallerType::CrossOriginIsolated;
+  }
+
+  return RTPCallerType::Normal;
+}